activator/src/main/java/org/pentaho/osgi/legacy/LegacyPluginExtenderFactory.java

```diff
@@ -0,0 +1,243 @@
+import org.osgi.framework.wiring.BundleWiring;
+import org.pentaho.di.core.exception.KettlePluginException;
+import org.pentaho.di.core.plugins.PluginInterface;
+import org.pentaho.di.core.plugins.PluginRegistry;
+import org.pentaho.di.core.plugins.PluginTypeInterface;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URL;
+import java.util.List;
+
+/**
+ * Factory class to instantiate services implementing interfaces known to OSGi using classes from a legacy plugin
+ */
+public class LegacyPluginExtenderFactory {
+  private final ClassLoader legacyBridgingClassloader;
+
+  /**
+   * Constructs the LegacyPluginExtenderFactory
+   *
+   * @param bundleContext the bundleContext
+   * @param pluginType    the pluginType's canonical name
+   * @param pluginId      the plugin id
+   * @throws KettlePluginException
+   */
+  public LegacyPluginExtenderFactory( BundleContext bundleContext, String pluginType, String pluginId )
+    throws KettlePluginException {
+    ClassLoader parentClassloader = getPluginClassloader( pluginType, pluginId );
+    if ( parentClassloader == null ) {
+      throw new KettlePluginException(
+        "Unable to get parent classloader for pluginType: " + pluginType + " and pluginId " + pluginId );
+    } else {
+      System.out.println( "Parent classloader: " + parentClassloader );
+    }
+    legacyBridgingClassloader = new LegacyPluginExtenderClassLoader( parentClassloader, bundleContext );
+  }
+
+  /**
+   * Gets the classloader for the specified plugin, checking once a second for it to come online (the timeout is
+   * deliberately long as the feature watcher will kill us after a while anyway
+   *
+   * @param pluginType the plugin type (Specified as a string so that we can get the classloader for plugin types OSGi
+   *                   doesn't know about)
+   * @param pluginId   the plugin id
+   * @return
+   * @throws KettlePluginException
+   * @throws InterruptedException
+   */
+  private static ClassLoader getPluginClassloader( String pluginType, String pluginId )
+    throws KettlePluginException {
+    Class<? extends PluginTypeInterface> pluginTypeInterface = null;
+    PluginRegistry pluginRegistry = PluginRegistry.getInstance();
+    int maxTries = 10000;
+    while ( maxTries-- > 0 ) {
+      if ( pluginTypeInterface == null ) {
+        for ( Class<? extends PluginTypeInterface> potentialPluginTypeInterface : pluginRegistry.getPluginTypes() ) {
+          if ( pluginType.equals( potentialPluginTypeInterface.getCanonicalName() ) ) {
+            pluginTypeInterface = potentialPluginTypeInterface;
+          }
+        }
+      }
+      PluginInterface plugin = pluginRegistry.getPlugin( pluginTypeInterface, pluginId );
+      if ( plugin != null ) {
+        return pluginRegistry.getClassLoader( plugin );
+      }
+      try {
+        Thread.sleep( 1000 );
```

Comment:
This will be replaced with wait() on the pluginRegistry. PluginRegistry will call notify() whenever a plugin is registered. At the very least this will aid someone in debugging deadlocks and be more performant.